### PR TITLE
WIP: Add rpm support for multiple distributions

### LINKED ISSUE #25304: Assign a distribution to yum
### Feature Description

Sometimes we need to compile RPM packages for multiple distributions ` (el7, el9, f34 ...)`  All packages are put together and sometimes a package does not work on a particular distribution, which may lead to incorrect installation of a package not compiled for that distribution.

图片

图片

图片

After uploading:

图片

### PR DESCRIPTION
**Pull request by ExplodingDragon**
**WIP**: The single distributions category is not well suited for complex environments and needs to be redesigned.

Closes #25304

The current rpm repository places all packages in the same repository, and different systems (el7,f34) may hit packages that do not belong to this distribution ( #25304 )  , which now supports grouping of rpm.

## :warning: BREAKING :warning:

Migrate old RPM packages to  `default` distribution, remove the old RPM upload API  and redirecting the download path to the default distribution.

图片

### CODE REVIEW

File: templates/package/content/rpm.tmpl [hunk 1]
```diff
@@ -18,9 +18,25 @@
...
+				<tr>
+					<td class="collapsing"><h5>{{.locale.Tr "packages.rpm.package.architectures"}}</h5></td>
+					<td>{{StringUtils.Join .Architectures ", "}}</td>
+				</tr>
+				{{if .PackageDescriptor.Metadata.Summary}}
+				<tr>
+					<td class="collapsing"><h5>{{.locale.Tr "packages.rpm.package.summary"}}</h5></td>
+					<td>{{.PackageDescriptor.Metadata.Summary}}</td>
+				</tr>
+				{{end}}
```
Comment (KN4CK3R) [inline templates/package/content/rpm.tmpl:46]: `.PackageDescriptor.Metadata.Description` is lost

Comment (ExplodingDragon) [reply templates/package/content/rpm.tmpl:46]: Already added.

图片

File: docs/content/doc/usage/packages/rpm.en-us.md [hunk 2]
```diff
@@ -5,11 +5,11 @@ slug: "rpm"
...
-  sidebar:
-    parent: "packages"
-    name: "RPM"
-    weight: 105
-    identifier: "rpm"
+sidebar:
+parent: "packages"
+name: "RPM"
+weight: 105
+identifier: "rpm"
```
Comment (lunny) [inline docs/content/doc/usage/packages/rpm.en-us.md]: The changes are not right.

File: routers/api/packages/api.go [hunk 3]
```diff
@@ -513,14 +513,14 @@ func CommonRoutes() *web.Route {
 			r.Get("/simple/{id}", pypi.PackageMetadata)
 		}, reqPackageAccess(perm.AccessModeRead))
 		r.Group("/rpm", func() {
-			r.Get(".repo", rpm.GetRepositoryConfig)
+			r.Get("/{distribution}.repo", rpm.GetRepositoryConfig)
```
Comment (lunny) [inline routers/api/packages/api.go]: Is this a break change?

File: routers/api/packages/api.go [hunk 4]
```diff
@@ -513,14 +513,23 @@ func CommonRoutes() *web.Route {
...
 			})
-			r.Get("/repodata/{filename}", rpm.GetRepositoryFile)
+			// new rpm
+			r.Get("/{distribution}.repo", rpm.GetRepositoryConfig)
+			r.Put("/{distribution}/upload", reqPackageAccess(perm.AccessModeWrite), rpm.UploadPackageFile)
+			r.Group("/{distribution}/package/{name}/{version}/{architecture}", func() {
+				r.Get("", rpm.DownloadPackageFile)
+				r.Delete("", reqPackageAccess(perm.AccessModeWrite), rpm.DeletePackageFile)
+			})
+			r.Get("/{distribution}/repodata/{filename}", rpm.GetRepositoryFile)
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]:
```suggestion
			r.Group("/{distribution}", func() {
				r.Get(".repo", rpm.GetRepositoryConfig)
				r.Put("/upload", reqPackageAccess(perm.AccessModeWrite), rpm.UploadPackageFile)
				r.Group("/package/{name}/{version}/{architecture}", func() {
					r.Get("", rpm.DownloadPackageFile)
					r.Delete("", reqPackageAccess(perm.AccessModeWrite), rpm.DeletePackageFile)
				})
				r.Get("/repodata/{filename}", rpm.GetRepositoryFile)
			})
```

File: routers/api/packages/api.go [hunk 5]
```diff
@@ -513,14 +513,23 @@ func CommonRoutes() *web.Route {
 			r.Get("/simple/{id}", pypi.PackageMetadata)
 		}, reqPackageAccess(perm.AccessModeRead))
 		r.Group("/rpm", func() {
+			// old rpm
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]: Only the old `GET` routes are needed. I suggest using a simple redirect with the default distribution. Then you don't need the `getDistribution` method.

File: routers/api/packages/api.go [hunk 6]
```diff
@@ -513,14 +513,27 @@ func CommonRoutes() *web.Route {
 			r.Get("/simple/{id}", pypi.PackageMetadata)
 		}, reqPackageAccess(perm.AccessModeRead))
 		r.Group("/rpm", func() {
-			r.Get(".repo", rpm.GetRepositoryConfig)
+			// old rpm
+			r.Get(".repo", func(ctx *context.Context) {
+				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm.repo", "/rpm/default.repo", 1), 302)
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]:
```suggestion
				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm.repo", "/rpm/default.repo", 1))
```

File: routers/api/packages/api.go [hunk 7]
```diff
@@ -513,14 +513,27 @@ func CommonRoutes() *web.Route {
...
 			r.Get("/repository.key", rpm.GetRepositoryKey)
-			r.Put("/upload", reqPackageAccess(perm.AccessModeWrite), rpm.UploadPackageFile)
-			r.Group("/package/{name}/{version}/{architecture}", func() {
-				r.Get("", rpm.DownloadPackageFile)
-				r.Delete("", reqPackageAccess(perm.AccessModeWrite), rpm.DeletePackageFile)
+			r.Get("/repodata/{filename}", func(ctx *context.Context) {
+				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm/repodata/", "/rpm/default/repodata/", 1), 302)
+			})
+			r.Get("/package/{name}/{version}/{architecture}", func(ctx *context.Context) {
+				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm/package/", "/rpm/default/package/", 1), 302)
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]:
```suggestion
			r.Get("/repodata/{filename}", func(ctx *context.Context) {
				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm/repodata/", "/rpm/default/repodata/", 1))
			})
			r.Get("/package/{name}/{version}/{architecture}", func(ctx *context.Context) {
				ctx.Redirect(setting.AppSubURL+strings.Replace(ctx.Base.Req.URL.Path, "/rpm/package/", "/rpm/default/package/", 1))
```

File: routers/api/packages/rpm/rpm.go [hunk 8]
```diff
@@ -33,11 +33,12 @@ func apiError(ctx *context.Context, status int, obj any) {
 
 // https://dnf.readthedocs.io/en/latest/conf_ref.html
 func GetRepositoryConfig(ctx *context.Context) {
+	distribution := strings.TrimSpace(ctx.Params("distribution"))
```
Comment (KN4CK3R) [inline routers/api/packages/rpm/rpm.go]: I think you don't need the `strings.TrimSpace` because we don't trim any of the other parameters.

**Review by KN4CK3R — COMMENTED** [review]
About the "breaking change": I don't think we need the redirect. The RPM registry was not released until now. But then this PR needs to be part of 1.20.

**Comment by ExplodingDragon** [comment]
@KN4CK3R Sorry, this pr I designed for 1.21, if this PR will be merged to 1.20 is it possible to remove all the migration code?

**Comment by ExplodingDragon** [comment]
> About the "breaking change": I don't think we need the redirect. The RPM registry was not released until now. But then this PR needs to be part of 1.20.

Then I would remove all redirect codes, `models/migrations/v1_21/v265.go` needs to be removed if merged into 1.20.

**Comment by delvh** [comment]
That's not possible.
So, it's indeed a breaking change.

**Comment by ExplodingDragon** [comment]
> That's not possible. So, it's indeed a breaking change.

Then I will keep the `GET` redirection.

If a user is migrating from 1.20 to 1.21 they may need to provide API migration instructions.

**Comment by KN4CK3R** [comment]
There was no non-beta version which contained the RPM registry so I would see this PR as bugfix for 1.20. I would keep the migration and remove the redirect. Just tell the early RPM users to update their repo file.

**Comment by ExplodingDragon** [comment]
Relevant test cases have been added.

**Comment by ExplodingDragon** [comment]
Currently 1.20 has been released. There is no need to consider merging to 1.20.

**Comment by ExplodingDragon** [comment]
> There was no non-beta version which contained the RPM registry so I would see this PR as bugfix for 1.20. I would keep the migration and remove the redirect. Just tell the early RPM users to update their repo file.

Currently 1.20 has been released.

**Comment by ExplodingDragon** [comment]
@KN4CK3R Given that 1.20 has been released, I rolled back the old path.

File: routers/api/packages/rpm/rpm.go [hunk 9]
```diff
@@ -69,7 +71,8 @@ func GetRepositoryFile(ctx *context.Context) {
 		ctx,
 		pv,
 		&packages_service.PackageFileInfo{
-			Filename: ctx.Params("filename"),
+			Filename:     ctx.Params("filename"),
+			CompositeKey: distribution,
```
Comment (lunny) [inline routers/api/packages/rpm/rpm.go]: Should we check `distribution` whether it's valid value?

File: routers/web/user/package.go [hunk 10]
```diff
@@ -213,6 +214,21 @@ func ViewPackageVersion(ctx *context.Context) {
 		ctx.Data["Distributions"] = distributions.Values()
 		ctx.Data["Components"] = components.Values()
 		ctx.Data["Architectures"] = architectures.Values()
+	case packages_model.TypeRpm:
```
Comment (lunny) [inline routers/web/user/package.go:223]: Is this a bug of previous code?

**Review by lunny — APPROVED** [review]

File: routers/api/packages/rpm/rpm.go [hunk 11]
```diff
@@ -59,6 +60,7 @@ func GetRepositoryKey(ctx *context.Context) {
 
 // Gets a pre-generated repository metadata file
 func GetRepositoryFile(ctx *context.Context) {
+	distribution := ctx.Params("distribution")
```
Comment (KN4CK3R) [inline routers/api/packages/rpm/rpm.go]: Could you move the `ctx.Params("distribution")` into the `PackageFileInfo` constructor. We don't need this variable.

File: routers/api/packages/rpm/rpm.go [hunk 12]
```diff
@@ -85,6 +88,7 @@ func GetRepositoryFile(ctx *context.Context) {
 }
 
 func UploadPackageFile(ctx *context.Context) {
+	distribution := ctx.Params("distribution")
```
Comment (KN4CK3R) [inline routers/api/packages/rpm/rpm.go]: Here you could move it above `CreatePackageOrAddFileToExisting`. Then initialization and usage are at the same place.

File: routers/api/packages/rpm/rpm.go [hunk 13]
```diff
@@ -168,7 +175,7 @@ func UploadPackageFile(ctx *context.Context) {
 func DownloadPackageFile(ctx *context.Context) {
 	name := ctx.Params("name")
 	version := ctx.Params("version")
-
+	distribution := ctx.Params("distribution")
```
Comment (KN4CK3R) [inline routers/api/packages/rpm/rpm.go]: Same, not needed.

File: routers/web/user/package.go (hunk 10 above)
Comment (KN4CK3R) [reply routers/web/user/package.go:223]: No, there was no need to display these infos.

File: tests/integration/api_packages_rpm_test.go [hunk 14]
```diff
@@ -27,14 +27,33 @@ import (
...
+	}
+
+	pkgsInfo := []PkgInfo{
+		{
+			packageName:         "gitea-test",
+			packageVersion:      "1.0.2-1",
+			packageArchitecture: "x86_64",
+		},
+		{
+			packageName:         "minio-client",
```
Comment (KN4CK3R) [inline tests/integration/api_packages_rpm_test.go]: Why do we need two packages to test with? Aren't the two distributions enough?

**Pull request closed without merging**